src/prognostic_equations/edmfx_closures.jl

```diff
@@ -206,78 +226,193 @@ function mixing_length(
+    l_W_base = vkc * l_z / l_W_denom
+
+    if obukhov_length < FT(0) # Unstable case
+        obukhov_len_safe = min(obukhov_length, -eps_FT) # Ensure L < 0
+        zeta = l_z / obukhov_len_safe # Stability parameter zeta = z/L (<0)
+
+        # Calculate MOST term (1 - b_m * zeta)
+        # Since zeta is negative, this term is > 1
+        inner_term = FT(1) - most_b_m * zeta
+
+        # Numerical safety check – by theory the value is ≥ 1.
+        inner_term_safe = max(inner_term, eps_FT)
+
+        # Unstable-regime correction factor:
+        #     (1 − b_m ζ)^(1/4) = φ_m⁻¹,
+        # where φ_m is the Businger stability function φ_m = (1 − b_m ζ)^(-1/4).
+        stability_correction = sqrt(sqrt(inner_term_safe))
+        l_W = l_W_base * stability_correction
+
+    else # Neutral or stable case
+        # Ensure L > 0 for Monin-Obukhov length
+        obukhov_len_safe_stable = max(obukhov_length, eps_FT)
+        zeta = l_z / obukhov_len_safe_stable # zeta >= 0
+
+        # Stable/neutral-regime correction after Gryanik (2020):
+        #     φ_m = 1 + a_m ζ / (1 + g_m ζ)^(2/3),
+        # a nonlinear refinement to the Businger formulation.
+        phi_m_denom_term = (FT(1) + most_g_m * zeta)
+        # Guard against a negative base in the fractional power
+        # (theoretically impossible for ζ ≥ 0 and g_m > 0, retained for robustness).
+        phi_m_denom_cubed_sqrt = cbrt(phi_m_denom_term)
+        phi_m_denom =
+            max(phi_m_denom_cubed_sqrt * phi_m_denom_cubed_sqrt, eps_FT) # (val)^(2/3)
+
+        phi_m = FT(1) + (most_a_m * zeta) / phi_m_denom
+
+        # Stable-regime correction factor: 1 / φ_m.
+        # phi_m should be >= 1 for stable/neutral
+        stability_correction = FT(1) / max(phi_m, eps_FT)
+
+        # Apply the correction factor
+        l_W = l_W_base * stability_correction
     end
-
-    # compute l_TKE - the production-dissipation balanced length scale
-    a_pd = c_m * (2 * ᶜstrain_rate_norm - ᶜlinear_buoygrad / ᶜPr) * sqrt(ᶜtke)
-    # Dissipation term
-    c_neg = c_d * ᶜtke * sqrt(ᶜtke)
-    if abs(a_pd) > eps(FT) && 4 * a_pd * c_neg > -(ᶜtke_exch * ᶜtke_exch)
-        l_TKE = max(
-            -(ᶜtke_exch / 2 / a_pd) +
-            sqrt(ᶜtke_exch * ᶜtke_exch + 4 * a_pd * c_neg) / 2 / a_pd,
-            0,
-        )
-    elseif abs(a_pd) < eps(FT) && abs(ᶜtke_exch) > eps(FT)
-        l_TKE = c_neg / ᶜtke_exch
-    else
-        l_TKE = FT(0)
+    l_W = max(l_W, FT(0)) # Ensure non-negative
+
+    # --- l_TKE: TKE production-dissipation balance scale ---
+    tke_pos = max(ᶜtke, FT(0)) # Ensure TKE is not negative
+    sqrt_tke_pos = sqrt(tke_pos)
+
+    # Net production of TKE from shear and buoyancy is approximated by
+    #     (S² − N²/Pr_t) · √TKE · l,
+    # where S² denotes shear production and N² denotes buoyancy production.
+    # The factor below corresponds to that production term normalised by l.
+    # Ensure ᶜPr is strictly positive.
+    pr_safe = max(ᶜPr, eps_FT)
```

Comment:
Is `Pr` number coming from `turbulent_prandtl_number`? If yes, then its already limited between `eps(FT)` and `Pr_max`

Reply:
yes - then I'll remove it